interfaces/thirdparty/cp2k.py

```diff
@@ -140,25 +142,74 @@ def get_runtime(self):
+        """Return the Timings from the End of the Output File"""
+        chunk = self.get_output_chunk(begin="SUBROUTINE",end="-"*70)
+        ret = {}
+        for line in chunk[1:]:
+            l0, *l1 = line.split()
+            ret[l0] = l1
+        return ret
+
+
+    def _get_energy_type(self, search='Total', index=-1, unit='a.u.'):
+        s = self.grep_output(search+' energy:')[index]
+        if not isinstance(index, slice):
+            return Units.convert(float(s.split()[-1]), 'a.u.', unit)
         else:
-            select = -1
-        s = self.grep_output(search + ' energy:')[select].split()[-1]
-        return float(s)
+            return [ Units.convert(float(x.split()[-1]), 'a.u.', unit) for x in s ]
+
+    def get_energy(self, index=-1, unit='a.u.'):
+        """Returns 'Total energy:' from the output file.
+
+        Set ``index`` to choose the n-th occurence of the Charges in the output, e.g. to choose an optimization step.
+        Also supports slices.
+        Defaults to the last occurence.
+        """
+        return self._get_energy_type('Total', index=index, unit=unit)
+
+    def get_dispersion(self, index=-1, unit='a.u.'):
+        """Returns 'Dispersion energy:' from the output file.
+
+        Set ``index`` to choose the n-th occurence of the Charges in the output, e.g. to choose an optimization step.
+        Also supports slices.
+        Defaults to the last occurence.
+        """
+        return self._get_energy_type('Dispersion', index=index, unit=unit)
 
-    def get_energy(self, index=0):
-        """Return last occurence of 'Total energy:' in the output."""
-        return self._get_energy_type('Total', index=index)
+    def get_forces(self, file=None, index=-1):
+        """Returns list of ndarrays with forces for each atom.
 
-    def get_dispersion(self, index=0):
-        """Return last occurence of 'Dispersion energy:' in the output."""
-        return self._get_energy_type('Dispersion', index=index)
+        Set ``file`` to use other than the main output file.
+
+        Set ``index`` to choose the n-th occurence of the Charges in the output, e.g. to choose an optimization step.
+        Set to *None* to return all as a list.
+        Defaults to the last occurence.
+        """
+        if not file:
+            file = self.job._filename('out')
+        searchBegin = "ATOMIC FORCES in"
+        searchEnd = "SUM OF ATOMIC FORCES"
+        n = len(self.grep_file(file, searchBegin))
+        match = self._idx_to_match(n, index)
+        block = self.get_file_chunk(file, begin=searchBegin, end=searchEnd, match=match)
+        ret = []
+        for line in block:
+            line = line.strip().split()
+            if len(line) == 0:
+                continue
+            #new forces block
+            if line[0] == '#':
+                ret.append([])
+                continue
+            ret[-1].append(line[-3:])
+        return [np.array(item, dtype=float) for item in ret]
```

Comment:
I guess in almost all cases they will be the same length. But in any simulation without constant number of particles, the array length might differ between frames. Not sure if there is any calculation in CP2K that allows this (perhaps some embedded schemes?!), but I guess this way we are on the safe side...

Reply:
> I guess in almost all cases they will be the same length.

By the sound of it the this would entail the vast majority (all?) cases.

How about converting the whole thing into an `ndarray` via a `try` / `except` approach,
using a ragged array as fallback if necessary?

``` python
import warnings
import numpy as np

def _sequence_to_array(seq, dtype=float):
    """Convert the (nested) sequence *seq* into an array with the specified *dtype*.
    
    If *seq* is ragged sequence it will be converted into a ragged object array instead.
    """
    try:
        return np.array(seq, dtype=dtype)
    except ValueError as ex:
        warning = RuntimeWarning(f"Failed to coerce the forces into a single {np.dtype(dtype)} array;"
                                 "returning a ragged array instead")
        warning.__cause__ = ex
        warnings.warn(warning, stacklevel=3)
        
        # Create and return a ragged array, 
        # i.e. an object array whose elements are normal arrays
        ret = np.empty(len(seq), dtype=object)
        ret[:] = [np.array(item, dtype=dtype) for item in seq]
        return ret

class Cp2kResults:
    get_forces(...):
        ...
        return _sequence_to_array(ret)
```